src/llvm-demote-float16.cpp

```diff
@@ -43,13 +45,54 @@ INST_STATISTIC(FRem);
 INST_STATISTIC(FCmp);
 #undef INST_STATISTIC
 
+extern JuliaOJIT *jl_ExecutionEngine;
+
+Optional<bool> always_have_fp16() {
+#if defined(_CPU_X86_) || defined(_CPU_X86_64_)
+    // x86 doesn't support fp16
+    // TODO: update for sapphire rapids when it comes out
+    return false;
+#else
+    return {};
+#endif
+}
+
 namespace {
 
+bool have_fp16(Function &caller) {
+    auto unconditional = always_have_fp16();
+    if (unconditional.hasValue())
+        return unconditional.getValue();
+
+    Attribute FSAttr = caller.getFnAttribute("target-features");
+    StringRef FS =
+        FSAttr.isValid() ? FSAttr.getValueAsString() : jl_ExecutionEngine->getTargetFeatureString();
+#if defined _CPU_AARCH64_
+    if (FS.find("+fp16fml") != llvm::StringRef::npos || FS.find("+fullfp16") != llvm::StringRef::npos){
+        return true;
+    }
+#else
+    if (FS.find("+avx512fp16") != llvm::StringRef::npos){
```

Comment:
Note https://reviews.llvm.org/D107082 we are not there yet, but LLVM will support `_Float16` correctly on SSE2 and above.

Note that the LLVM PR also changes the ABI to match GCC12 and thus is going to break us in fun ways. I haven't found how `-fexcess-precision=16` is going to be implemented in LLVM,

Reply:
I just added this there because GCC was complaining about FS being unused. That part of the branch doesn't matter for now since x86 is considered as never having Float16 for now. 
SSE2 has f16C instructions which are just fast conversions which we might already use, I know we use them on aarch64 at least. The first native operations on float16 are the avx512 ones.